nemo/collections/nlp/models/language_modeling/megatron/bert/bert_model.py

```diff
@@ -154,6 +174,260 @@ def post_language_model_processing(
+            hidden_states = self.cross_attn_bda(self.training, self.config.bias_dropout_fusion)(
+                attention_output_with_bias, residual, self.hidden_dropout
+            )
+
+        # Residual connection.
+        residual = hidden_states
+
+        # Optional Layer norm post the cross-attention.
+        pre_mlp_layernorm_output = self.pre_mlp_layernorm(hidden_states)
+
+        # MLP.
+        mlp_output_with_bias = self.mlp(pre_mlp_layernorm_output)
+
+        # TODO: could we move `bias_dropout_add_exec_handler` itself
+        # inside the module provided in the `bias_dropout_add_spec` module?
+        with self.bias_dropout_add_exec_handler():
+            hidden_states = self.mlp_bda(self.training, self.config.bias_dropout_fusion)(
+                mlp_output_with_bias, residual, self.hidden_dropout
+            )
+
+        # Post-LN after MLP
+        hidden_states = self.post_mlp_layernorm(hidden_states)
+
+        # Jit compiled function creates 'view' tensor. This tensor
+        # potentially gets saved in the MPU checkpoint function context,
+        # which rejects view tensors. While making a viewless tensor here
+        # won't result in memory savings (like the data loader, or
+        # p2p_communication), it serves to document the origin of this
+        # 'view' tensor.
+        output = make_viewless_tensor(inp=hidden_states, requires_grad=hidden_states.requires_grad, keep_graph=True)
+
+        return output, context
+
+
+class TransformerBlockWithPostLNSupport(TransformerBlock):
+    def __init__(self, transformer_block_type='post_ln', *args, **kwargs):
+
+        super(TransformerBlockWithPostLNSupport, self).__init__(*args, **kwargs)
+        self.transformer_block_type = transformer_block_type
+        if self.transformer_block_type == 'post_ln':
+            self.initial_layernorm = FusedLayerNorm(
+                config=self.config, hidden_size=self.config.hidden_size, eps=self.config.layernorm_epsilon
+            )
+
+    def forward(
+        self,
+        hidden_states: Tensor,
+        attention_mask: Tensor,
+        context: Tensor = None,
+        context_mask: Tensor = None,
+        rotary_pos_emb: Tensor = None,
+        inference_params: InferenceParams = None,
+        packed_seq_params: PackedSeqParams = None,
+    ):
+        # hidden_states (float): [s, b, h]
+        # attention_mask (bool): [1, 1, s, s]
+
+        if not self.pre_process:
+            # See set_input_tensor()
+            hidden_states = self.input_tensor
+        if self.transformer_block_type == 'post_ln':
+            hidden_states = self.initial_layernorm(hidden_states)
+        return super(TransformerBlockWithPostLNSupport, self).forward(
+            hidden_states, attention_mask, context, context_mask, rotary_pos_emb, inference_params, packed_seq_params
+        )
+
+
+# For converting HF Bert checkpoints, we are adding this wrapper to support post layernorm
```

Comment:
This class is used for working with HF Bert Checkpoints. These checkpoints by default have post layer norm, while the vanilla mcore bert model does not support it.